webpack/webpack-workspace-resolve-plugin/src/WorkspaceLayoutCache.ts

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { sep as directorySeparator } from 'node:path';
+
+import { LookupByPath, type IPrefixMatch } from '@rushstack/lookup-by-path';
+
+/**
+ * Information about a local or installed npm package.
+ * @beta
+ */
+export interface ISerializedResolveContext {
+  /**
+   * The path to the root folder of this context.
+   * This path is normalized to use `/` as the separator and should not end with a trailing `/`.
+   */
+  root: string;
+  /**
+   * The name of this package. Used to inject a self-reference into the dependency map.
+   */
+  name: string;
+  /**
+   * Map of declared dependencies to the ordinal of the corresponding context.
+   */
+  deps: Record<string, number>;
+  /**
+   * Set of relative paths to nested `package.json` files within this context.
+   * These paths are normalized to use `/` as the separator and should not begin with a leading `./`.
+   */
+  dirInfoFiles?: string[];
+}
+
+/**
+ * The serialized form of the cache file. This file is expected to be generated by a separate tool from
+ * information known to the package manager. Namely, the dependency relationships between packages, and
+ * all the `package.json` files in the workspace (installed or local).
+ * @beta
+ */
+export interface IResolverCacheFile {
+  /**
+   * The ordered list of all contexts in the cache
+   */
+  contexts: ISerializedResolveContext[];
+}
+
+/**
+ * A context for resolving dependencies in a workspace.
+ * @beta
+ */
+export interface IResolveContext {
+  /**
+   * The absolute path to the root folder of this context
+   */
+  descriptionFileRoot: string;
+  /**
+   * Find the context that corresponds to a module specifier, when requested in the current context.
+   * @param request - The module specifier to resolve
+   */
+  findDependency(request: string): IPrefixMatch<IResolveContext> | undefined;
+}
+
+/**
+ * Options for creating a `WorkspaceLayoutCache`.
+ * @beta
+ */
+export interface IWorkspaceLayoutCacheOptions {
+  /**
+   * The root folder of the workspace. All paths in the cache file are assumed to be relative to this folder.
+   */
+  workspaceRoot: string;
+  /**
+   * The parsed cache data. File reading is left as an exercise for the caller.
+   */
+  cacheData: IResolverCacheFile;
+  /**
+   * The directory separator used in the `path` field of the resolver inputs.
+   * Will usually be `path.sep`.
+   */
+  resolverPathSeparator?: '/' | '\\';
+}
+
+function preservePath(path: string): string {
+  return path;
+}
+
+function backslashToSlash(path: string): string {
+  return path.replace(/\\/g, '/');
+}
+
+function slashToBackslash(path: string): string {
+  return path.replace(/\//g, '\\');
+}
```

Comment:
There are utility functions on `Path` from `node-core-library`.

Reply:
Which is a rather large npm dependency footprint for two trivial utility functions, that this package currently doesn't take.